### ISSUE TITLE
Can't implement batching rule for F.one_hot as written

### ISSUE THREAD

**Issue by Chillee** [facebookresearch/functorch#85]
We can't currently implement a batching rule for `F.one_hot`, even if we had batching rules for the constituent ops (`_local_scalar_dense`, `min`, `max`, `unsqueeze`, `scatter`).

In the C++ code, the code looks like this here (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Onehot.cpp#L38)
```
    Tensor ret = at::zeros(shape, self.options());
    ret.scatter_(-1, self.unsqueeze(-1), 1);
    return ret;
```
cc: @bdhirsh

**Comment by pedro-abreu** [comment]
Any ideas how to solve this? I've been using vmap and found this severely impacted my performance.

**Comment by Chillee** [comment]
@pedro-abreu We can prioritize fixing it - there's a couple different ways we can resolve it.

One of them is by decomposing it into a `zeros` + non-inplace scatter. We can do that.

**Comment by zou3519** [comment]
@pedro-abreu we've added a quick fix in the form of https://github.com/facebookresearch/functorch/pull/93. If you install functorch again using the latest main branch that should do the job. Please let us know if there are other batching rules we can help implement for your use case!

**Comment by pedro-abreu** [comment]
Thank you, I will test it soon, so you can close this issue.

**Comment by pedro-abreu** [comment]
My performance issues are fixed and batching works fine now :) Thank you!

**Comment by zou3519** [comment]
@pedro-abreu out of curiosity (and to collect some feedback), could you please tell us a little bit about what you're using functorch for?

**Comment by pedro-abreu** [comment]
@pedro-abreu trying to implement a differentiable top-k :)

**Comment by zou3519** [comment]
@pedro-abreu that sounds cool! Does that mean that [torch.topk](https://pytorch.org/docs/stable/generated/torch.topk.html) was not sufficient for your use case?

**Comment by pedro-abreu** [comment]
@pedro-abreu Yes, basically you cannot backpropagate through torch.topk. But I think from my ignorant perspective of the underlying codebase, is it possible to batch a custom autograd function with functorch? In JAX you can do a custom vjp by providing defvjp() with a forward and backward function and easily vmap the function where you called defvjp().

**Comment by zou3519** [comment]
> is it possible to batch a custom autograd function with functorch? In JAX you can do a custom vjp by providing defvjp() with a forward and backward function and easily vmap the function where you called defvjp() (not sure if the vjp in functorch allows for this).

Not yet, but it is something we're looking into

**Comment by pedro-abreu** [comment]
> > is it possible to batch a custom autograd function with functorch? In JAX you can do a custom vjp by providing defvjp() with a forward and backward function and easily vmap the function where you called defvjp() (not sure if the vjp in functorch allows for this).
> 
> Not yet, but it is something we're looking into

Is there any way I can help?

**Comment by zou3519** [comment]
@pedro-abreu to be fair, I don't know what actually happens when one calls vmap with a custom autograd function. So there is a small possibility that it actually does work. If you have the bandwidth to play around with and see what exactly happens right now when someone calls vmap with an autograd.Function that would be helpful

**Comment by zou3519** [comment]
grad transform sometimes does not work with autograd.Function, like here: https://gist.github.com/zou3519/73a6189e21561f6ef5b42874e8a4826f
